### PR TITLE
Preserve file mode when copying files out of a pod

### PR DESCRIPTION
**Pull request by evanphx**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This fixes the ability to inherit the file modes from the files within the pod that are copied. Most importantly, it makes sure that the execute bits are set properly and generally makes this code match what tar would do (namely, using a readonly mode if the source was readonly)

**Which issue(s) this PR fixes**:
Didn't open an issue, just fixed the problem directly.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
File modes are properly maintained when copying data out of pods with `kubectl cp`
```

### CODE REVIEW

**Comment by evanphx** [comment]
signed the CLA

**Review by neolit123 — COMMENTED** [review]
thank you for the PR @evanphx 
this change needs approval by SIG CLI.

/ok-to-test
/priority backlog

**Comment by evanphx** [comment]
@neolit123 Any thoughts on this?

**Comment by neolit123** [comment]
> @neolit123 Any thoughts on this?

try mentioning this PR in the slack channels #sig-cli or #pr-reviews

File: pkg/kubectl/cmd/cp/cp.go [hunk 1]
```diff
@@ -477,7 +477,7 @@ func (o *CopyOptions) untarAll(reader io.Reader, destDir, prefix string) error {
 				return err
 			}
 		} else {
-			outFile, err := os.Create(destFileName)
+			outFile, err := os.OpenFile(destFileName, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, mode)
```
Comment (liggitt) [inline pkg/kubectl/cmd/cp/cp.go]: don't we need to mask the mode to just permission bits by passing in `mode.Perm()`?

**Comment by liggitt** [comment]
@soltysh @tallclair note that this will allow creating executable files via `kubectl cp`. What `kubectl cp` currently does is not correct, but we've had enough issues with `cp` that I'm really wary of expanding its potential as an attack vector

File: pkg/kubectl/cmd/cp/cp.go (hunk 1 above)
Comment (apelisse) [reply pkg/kubectl/cmd/cp/cp.go]: if we preserve the mode, doesn't it mean we could technically open the file in a non-writable mode? I suspect `cp` does copy the file AND THEN changes the permissions.

Comment (apelisse) [reply pkg/kubectl/cmd/cp/cp.go]: Also, cp doesn't preserve modes by default AFAIK.

**Comment by evanphx** [comment]
I added this actually because I wanted to preserve the executable bits.

**Comment by evanphx** [comment]
What should `kubectl cp` do? I can add a flag, ala tar, to preserve executable bits.

File: pkg/kubectl/cmd/cp/cp.go (hunk 1 above)
Comment (evanphx) [reply pkg/kubectl/cmd/cp/cp.go]: `cp` preserve modes in that if you copy an executable it's still executable. Not sure if you mean something else?

Comment (apelisse) [reply pkg/kubectl/cmd/cp/cp.go]: No, looks like I was wrong :-/

**Comment by evanphx** [comment]
I'm happy to change this however y'all think is best, please let me know.

**Comment by soggiest** [comment]
Hello! Code Freeze is just about on us. We'll be entering into Freeze tomorrow, May 31st. Is this still planned to land in 1.15?

**Comment by apelisse** [comment]
> What should kubectl cp do? I can add a flag, ala tar, to preserve executable bits.

I think I'm fine with that, what do you think @liggitt @soltysh

**Comment by liggitt** [comment]
given the repeated past vulnerabilities in `kubectl cp` where files from malicious tar files sent by containers could escape the target directory, I don't think making files executable is a good idea. 

@kubernetes/sig-cli-maintainers should consider one or more of the following options:
* moving `kubectl cp` to an optional plugin (reduces attack surface for users that don't care about it, allows independent release of the command)
* rewriting `kubectl cp` to use `tar` directly instead of a custom tar file extractor. most tar implementations are hardened to avoid extracting files outside the target directory (via absolute paths, or relative backsteps, or symlinks)
* deprecating/removing `kubectl cp` in favor of direct use of `kubectl exec` and `tar`

**Comment by xmudrii** [comment]
@evanphx @liggitt Hello! I'm bug triage lead for the 1.16 release cycle and considering this PR has not been updated for a long time, I'd like to check what's the status of this PR. The code freeze is starting 29th August (about 2.5 weeks from now) and while there is plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.16, is it still planned for this release?

**Comment by evanphx** [comment]
@liggitt Perhaps since you're concerned about setting executable, exposing it as a configuration option to honor the executable bit? Happy to add that to this PR.

**Comment by liggitt** [comment]
the history of `kubectl cp` with respect to secure extraction of files is not good (c.f. [CVE-2018-1002100](https://github.com/kubernetes/kubernetes/issues/61297), [CVE-2019-1002101](https://github.com/kubernetes/kubernetes/pull/75037), [CVE-2019-11246](https://github.com/kubernetes/kubernetes/pull/76788), [CVE-2019-11249](https://github.com/kubernetes/kubernetes/issues/80984))

even if the option is flag-gated, the capability would still have escalated the severity of those issues significantly if the placed files could be made executable.

I'd prefer we document how to use `kubectl exec` in combination with `tar` to accomplish advanced use cases like this. @tallclair looked into that, I believe, and might be able to shepherd some docs changes to that effect.

**Comment by tallclair** [comment]
The basic command to copy from a pod using exec & tar is:
```
kubectl exec ${POD} -- tar cf - ${FILE} | tar xf -
```
Of course you can use any typical arguments to tar, in this case you might be interested in `--preserve --same-owner`.

More examples here: https://gist.github.com/tallclair/9217e2694b5fdf27b55d6bd1fda01b53

**Comment by soltysh** [comment]
I agree with @liggitt here, we (as in sig-cli) need to reconsider `kubectl cp` implementation. The current causes more harm than good to both users and maintainers. I don't think that moving this to an optional plugin is a good approach, I'd prefer the rewrite option. I'll rise this problem during next sig-cli call (Wed, Aug 28th).

**Comment by liggitt** [comment]
Discussed in sig-cli this week. The immediate priority is to resolve symlink handling issues in https://github.com/kubernetes/kubernetes/pull/82143 before considering expanding the surface area to include things like wildcard support and executable-permission preservation.

**Comment by evanphx** [comment]
I can rebase this but it sounded like before I should not bother because cp
is going to be removed?

**Comment by atomd** [comment]
tx

**Pull request closed without merging**